### PR TITLE
Update docs such that they can be used in knative/docs

### LINKED ISSUE #193: Create reference docs
## Expected Behavior
We should have 3 kinds of docs:

* Tutorials
* References
* Concepts

For example Concourse has:
* https://concoursetutorial.com/ <-- tutorials
* https://concourse-ci.org/reference.html <-- reference
* https://concourse-ci.org/concepts.html <-- concepts

References should go into detail about what can and can't be done with each of our CRDs.

Bonus: we should have some guidelines in our CONTRIBUTING.md about which docs belong where.

## Actual Behavior

We have:

* Concepts in docs/concepts.md
* Tutorials will be added in #22 and kind of exist in docs/using.md but need some more love

But we have no reference docs

### LINKED ISSUE #304: Update `knative/docs` with build-pipeline docs
## Expected Behavior
In version 0.3 of `knative`, we'd like to release `build-pipeline`, which means that the docs repo docs on Build (https://github.com/knative/docs/tree/master/build) should talk about `build-pipeline`.

I propose that we do this:
* Add all of the markdown files at https://github.com/knative/docs/tree/master/build to this repo
* Update those docs to have the corresponding info in a post Build -> Task world
* As part of our official versioned releases (relevant to #72) we update `knative/docs` with the corresponding version of these docs from the repo

This way:
* `knative/build-pipeline` contains docs which are versioned alongside the code
* `knative/docs` contains the docs as of the most recent release

## Actual Behavior

It looks like the docs at https://github.com/knative/docs/tree/master/build are updated manually and don't have a clear process that ties them to the build + buildtemplate repos (@ImJasonH correct me if I'm wrong!)

## Additional Info
Blocked by the epic #147

### PR DESCRIPTION
**Pull request by bobcatfish**
In version 0.4 of knative, we'd like to release build-pipeline, which
means that the docs repo (https://github.com/knative/docs) should talk
about build-pipeline.

I think it's important to make sure that docs are updated with the code,
so I propose we have living versions of these docs in this repo, and
every time we release, we copy the docs into the knative/docs repo, so
that this repo has the correct docs at HEAD and knative/docs has the
correct docs as of the latest release. (How we want to deal with docs
for previous releases will have to be another story!)

In this PR, I've gone through [the knative docs for
build](https://github.com/knative/docs/tree/master/build) and shaped our
existing docs to be similar, which meant:

* Taking content out of using.md and splitting it into a file for each
  type of CRD
* Adding reference docs for the CRD fields (fixes #193)
* Most of concepts.md and some of using.md became the README file which
  lives at the top level of the `docs` dir here and hopefully the new
  `pipelines` dir in knative/docs

This also meant gaining a lot of docs by taking existing Build docs and
updating them to talk about pipelines (e.g. docs about auth)

Fixes #304

Reviewers: I've probably missed some link, some references to Build, and some k8s standards. Your nits are very welcome!

(After we finish and merge this PR, I'll open a PR in knative/docs to add these docs there. If they request doc changes I'll backport them here)

### CODE REVIEW

File: docs/pipelines.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  - [`metadata`][kubernetes-overview] - Specifies data to uniquely identify the
+    `Pipeline` resource object, for example a `name`.
+  - [`spec`][kubernetes-overview] - Specifies the configuration information for
+    your `Pipeline` resource object. In order for a `Pipeline` to do anything, the
+    spec must include:
+    - [`tasks`](#pipeline-tasks) - Specifies which `Tasks` to run and how to run them
+- Optional:
+  - [`resources`](#declared-resources) - Specifies which [`PipelineResources`](resources.md)
+    of which types the `Pipeline` will be using in its [Tasks](#pipeline-tasks)
+  - [`timeout`](#timeout) - Specifies timeout after which the `Pipeline` will fail.
```
Comment (pivotal-nader-ziada) [inline docs/pipelines.md]: this link doesn't go anywhere

File: docs/pipelines.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+At a minimum, this declaration must include a reference to the `Task`:
+
+```yaml
+  tasks:
+    - name: build-the-image
+      taskRef:
+        name: build-push
+```
+
+[Declared `PipelineResources`](#declared-resources) can be given to `Task`s in the `Pipeline` as
```
Comment (pivotal-nader-ziada) [inline docs/pipelines.md]: This link doesn't go anywhere

File: docs/pipelines.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            resource: my-repo
+        outputs:
+          - name: image
+            resource: my-image
+```
+
+The resource `my-image` is expected to be given to the `deploy-app` `Task` from
+the `build-app` `Task`. This means that the `PipelineResource` `my-image` must
+also be declared as an output of `build-app`.
+
```
Comment (pivotal-nader-ziada) [inline docs/pipelines.md]: is that the same as `build-the-image` in the example above?

File: docs/pipelines.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+      - name: my-image
+        from:
+          - build-app
+```
+
+The resource `my-image` is expected to be given to the `deploy-app` `Task` from
+the `build-app` `Task`. This means that the `PipelineResource` `my-image` must
+also be declared as an output of `build-app`.
+
+For implementation details, see [the developer docs](docs/developers/README.md).
```
Comment (pivotal-nader-ziada) [inline docs/pipelines.md]: this link gives a `404` 😲

File: docs/tasks.md [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+- Optional:
+  - [`inputs`](#inputs) - Specifies parameters and [`PipelineResources`](resources.md)
+    needed by your `Task`
+  - [`outputs`](#outputs) - Specifies [`PipelineResources`](resources.md) needed by your `Task`
+  - [`serviceAccount`](#service-account) - Specifies a `ServiceAccount`
+    resource object that enables your build to run with the defined
+    authentication information.
+  - [`volumes`](#volumes) - Specifies one or more volumes that you want to make
+    available to your build.
+  - [`timeout`](#timeout) - Specifies timeout after which the `Pipeline` will fail.
```
Comment (pivotal-nader-ziada) [inline docs/tasks.md]: This link doesn't go anywhere

File: docs/tasks.md [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+  - [`spec`][kubernetes-overview] - Specifies the configuration information for
+    your `Task` resource object. `Task` steps must be defined through either of
+    the following fields:
+    - [`steps`](#steps) - Specifies one or more container images that you want
+      to run in your `Task`.
+- Optional:
+  - [`inputs`](#inputs) - Specifies parameters and [`PipelineResources`](resources.md)
+    needed by your `Task`
+  - [`outputs`](#outputs) - Specifies [`PipelineResources`](resources.md) needed by your `Task`
+  - [`serviceAccount`](#service-account) - Specifies a `ServiceAccount`
```
Comment (pivotal-nader-ziada) [inline docs/tasks.md]: This link doesn't go anywhere

File: docs/tasks.md [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+  from the top of the configuration file.
+- Each container image runs until completion or until the first failure is
+  detected.
+
+### Inputs
+
+A `Task` can declare the inputs it needs, which can be either or both of:
+
+* [`parameters`](#parameters)
+* [`input resources](#input-resources)
```
Comment (pivotal-nader-ziada) [inline docs/tasks.md]: extra character before the word input

File: docs/tasks.md [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+${inputs.params.<name>}
+```
+
+## Examples
+
+Use these code snippets to help you understand how to define your `Tasks`.
+
+- [Example of image building and pushing](#example-task)
+- [Mounting extra volumes](#using-an-extra-volume)
+- [Authenticating with `ServiceAccount`](#using-a-serviceaccount)
```
Comment (pivotal-nader-ziada) [inline docs/tasks.md]: this link doesn't go anywhere

File: docs/container-contract.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+steps:
+  - image: ubuntu
+    command: ["/bin/bash"]
+    args: ["-c", "echo hello $FOO"]
+    env:
+      - name: "FOO"
+        value: "world"
+```
+
+_See [the installation guide](installing.md) if you would like to
```
Comment (pivotal-nader-ziada) [inline docs/container-contract.md]: both links here give `404`

File: docs/pipelines.md (hunk 2 above)
Comment (bobcatfish) [reply docs/pipelines.md]: hmmm I think this one works 🤔

File: docs/pipelines.md (hunk 4 above)
Comment (bobcatfish) [reply docs/pipelines.md]: whoops!

File: docs/pipelines.md (hunk 1 above)
Comment (bobcatfish) [reply docs/pipelines.md]: 😅

File: docs/pipelines.md (hunk 3 above)
Comment (bobcatfish) [reply docs/pipelines.md]: whoa this block seems completely out of place 😊

File: docs/pipelines.md (hunk 4 above)
Comment (bobcatfish) [reply docs/pipelines.md]: im actually gonna remove this b/c the developer dir won't be copied over to knative/docs...

File: docs/tasks.md (hunk 6 above)
Comment (bobcatfish) [reply docs/tasks.md]: whoops, thats only in taskrun! 😇

File: docs/tasks.md (hunk 7 above)
Comment (bobcatfish) [reply docs/tasks.md]: 😅

File: docs/tasks.md (hunk 8 above)
Comment (bobcatfish) [reply docs/tasks.md]: 😅

File: docs/container-contract.md (hunk 9 above)
Comment (bobcatfish) [reply docs/container-contract.md]: oh yeah, we won't have that till we have #385 🤔

**Comment by bobcatfish** [comment]
Thanks for noticing all those details @pivotal-nader-ziada !! Ready for another look :D

**Comment by pivotal-nader-ziada** [comment]
looks good 👍

/lgtm

**Pull request merged**